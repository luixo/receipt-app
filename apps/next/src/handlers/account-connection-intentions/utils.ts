import * as trpc from "@trpc/server";

import { Database } from "next-app/db";
import { ACCOUNT_CONNECTIONS_INTENTIONS } from "next-app/db/consts";
import { AccountsId, UsersId } from "next-app/db/models";
import { getUserById } from "next-app/handlers/users/utils";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: string,
	asUserId: UsersId
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select("id")
		.where("email", "=", toEmail)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Account with email ${toEmail} does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId)
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new trpc.TRPCError({
			code: "CONFLICT",
			message: `Account with email ${toEmail} is already connected to user ${connectedUser.name}.`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where("accountId", "=", targetAccount.id)
		.where("targetAccountId", "=", fromAccountId)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where("ownerAccountId", "=", targetAccount.id)
				.where("id", "=", viceVersaIntention.userId)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where("ownerAccountId", "=", fromAccountId)
				.where("id", "=", asUserId)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where("accountId", "=", targetAccount.id)
				.where("targetAccountId", "=", fromAccountId)
				.executeTakeFirst();
		});
		return {
			id: targetAccount.id,
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				created: new Date(),
			})
			.executeTakeFirst();
		return {
			id: targetAccount.id,
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where("accountId", "=", fromAccountId)
				.where("targetAccountId", "=", targetAccount.id)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await getUserById(
				database,
				existingIntention.userId,
				["name"]
			);
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to ${toEmail} as user ${
					existingUser!.name
				}.`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user ${user.name}.`,
			});
		}
		throw e;
	}
};

export const removeIntention = async (
	database: Database,
	intention:
		| {
				targetAccountId: AccountsId;
				accountId: AccountsId;
		  }
		| undefined,
	intentionType: string
) => {
	if (!intention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Intention ${intentionType} does not exist.`,
		});
	}
	await database
		.deleteFrom("accountConnectionsIntentions")
		.where("accountId", "=", intention.accountId)
		.where("targetAccountId", "=", intention.targetAccountId)
		.execute();
};
