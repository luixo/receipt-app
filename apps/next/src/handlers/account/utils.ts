import { Selection } from "kysely";
import { Database, ReceiptsSelectExpression, ReceiptsDatabase } from "../../db";
import { AccountsId } from "../../db/models";

export const getAccountById = <SE extends ReceiptsSelectExpression<"accounts">>(
	database: Database,
	id: AccountsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("id", "=", id)
		.executeTakeFirst();
};

export const getAccountByEmail = <
	SE extends ReceiptsSelectExpression<"accounts">
>(
	database: Database,
	email: string,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("email", "=", email)
		.executeTakeFirst();
};
