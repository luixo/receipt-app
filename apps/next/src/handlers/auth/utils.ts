import * as trpc from "@trpc/server";

import { DAY } from "app/utils/time";
import type { Database } from "next-app/db";
import type { AccountsId, SessionsSessionId } from "next-app/db/models";
import { generateConfirmEmailEmail } from "next-app/email/utils";
import { getUuid } from "next-app/utils/crypto";
import { getEmailClient, isEmailServiceActive } from "next-app/utils/email";

// How long a session should last
const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long before the session expires should we update the expiration date
const SESSION_SHOULD_UPDATE_DURATION = 28 * DAY;

const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);
export const shouldUpdateExpirationDate = (expirationTimestamp: Date) =>
	expirationTimestamp.valueOf() - Date.now() < SESSION_SHOULD_UPDATE_DURATION;

export const createAuthorizationSession = async (
	database: Database,
	accountId: AccountsId,
) => {
	const uuid = getUuid();
	const expirationDate = getExpirationDate();
	await database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const updateAuthorizationSession = async (
	database: Database,
	authToken: SessionsSessionId,
) => {
	database
		.updateTable("sessions")
		.set({ expirationTimestamp: getExpirationDate() })
		.where("sessionId", "=", authToken)
		.executeTakeFirst();
};

export const removeAuthorizationSession = async (
	database: Database,
	sessionId: SessionsSessionId,
) => {
	await database
		.deleteFrom("sessions")
		.where("sessionId", "=", sessionId)
		.executeTakeFirst();
};

export const sendVerificationEmail = async (email: string, token: string) => {
	if (!isEmailServiceActive()) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient().send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(token),
		});
	} catch (e) {
		throw new trpc.TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
