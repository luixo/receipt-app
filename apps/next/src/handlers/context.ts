import * as trpcNext from "@trpc/server/adapters/next";
import { NextApiRequest, NextApiResponse } from "next";
import { Logger, logger } from "../utils/logger";

export type Context = {
	req: NextApiRequest;
	res: NextApiResponse;
	logger: Logger;
	debug: boolean;
};

export const createContext = (
	opts: trpcNext.CreateNextContextOptions
): Context => ({
	req: opts.req,
	res: opts.res,
	logger,
	debug: Boolean(opts.req.headers.debug),
});
