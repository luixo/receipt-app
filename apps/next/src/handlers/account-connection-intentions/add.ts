import * as trpc from "@trpc/server";
import { z } from "zod";

import { emailSchema } from "app/utils/validation";
import { getDatabase } from "next-app/db";
import { addConnectionIntention } from "next-app/handlers/account-connection-intentions/utils";
import { authProcedure } from "next-app/handlers/trpc";
import { getUserById } from "next-app/handlers/users/utils";
import { userIdSchema } from "next-app/handlers/validation";

export const procedure = authProcedure
	.input(
		z.strictObject({
			userId: userIdSchema,
			email: emailSchema,
		})
	)
	.mutation(async ({ input, ctx }) => {
		const database = getDatabase(ctx);
		const user = await getUserById(database, input.userId, [
			"id",
			"name",
			"connectedAccountId",
			"ownerAccountId",
		]);
		if (!user) {
			throw new trpc.TRPCError({
				code: "NOT_FOUND",
				message: `User ${input.userId} does not exist.`,
			});
		}
		if (user.ownerAccountId !== ctx.auth.accountId) {
			throw new trpc.TRPCError({
				code: "FORBIDDEN",
				message: `User ${input.userId} is not owned by ${ctx.auth.accountId}.`,
			});
		}
		if (user.connectedAccountId) {
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `User ${input.userId} is already connected to an account.`,
			});
		}
		return addConnectionIntention(
			database,
			ctx.auth.accountId,
			user,
			input.email,
			input.userId
		);
	});
