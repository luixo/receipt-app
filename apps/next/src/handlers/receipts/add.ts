import { v4 } from "uuid";
import { z } from "zod";

import { receiptNameSchema } from "app/utils/validation";
import { getDatabase } from "next-app/db";
import { addReceiptParticipants } from "next-app/handlers/receipt-participants/utils";
import { authProcedure } from "next-app/handlers/trpc";
import { currencySchema, userIdSchema } from "next-app/handlers/validation";

export const procedure = authProcedure
	.input(
		z.strictObject({
			name: receiptNameSchema,
			currency: currencySchema,
			userIds: z.array(userIdSchema).optional(),
			issued: z.date(),
		})
	)
	.mutation(async ({ input, ctx }) => {
		const id = v4();
		const database = getDatabase(ctx);
		await database.transaction().execute(async (tx) => {
			await tx
				.insertInto("receipts")
				.values({
					id,
					name: input.name,
					currency: input.currency,
					created: new Date(),
					issued: input.issued,
					ownerAccountId: ctx.auth.accountId,
				})
				.executeTakeFirst();
			if (input.userIds) {
				await addReceiptParticipants(
					tx,
					id,
					ctx.auth.accountId,
					input.userIds.map((userId) => [userId, "editor"])
				);
			}
		});
		return id;
	});
