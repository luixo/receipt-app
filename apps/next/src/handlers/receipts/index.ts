import * as trpc from "@trpc/server";
import { router as previewsRouter } from "./previews";
import { router as itemsRouter } from "./items";
import { AuthorizedContext } from "../context";

export const router = trpc
	.router<AuthorizedContext>()
	.merge(previewsRouter)
	.merge(itemsRouter);
