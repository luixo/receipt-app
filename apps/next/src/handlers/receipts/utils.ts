import { Selection } from "kysely";
import { z } from "zod";

import {
	ReceiptsSelectExpression,
	ReceiptsDatabase,
	Database,
} from "next-app/db";
import { AccountsId, ReceiptsId } from "next-app/db/models";
import { roleSchema } from "next-app/handlers/validation";

export type Role = z.infer<typeof roleSchema>;

export const getAccessRole = async (
	database: Database,
	receipt: { ownerAccountId: string; id: string },
	accountId: string
): Promise<Role | undefined> => {
	if (receipt.ownerAccountId === accountId) {
		return "owner";
	}
	const participant = await database
		.selectFrom("receiptParticipants")
		.innerJoin("users", (jb) =>
			jb.onRef("users.id", "=", "receiptParticipants.userId")
		)
		.innerJoin("accounts", (jb) =>
			jb.onRef("accounts.id", "=", "users.connectedAccountId")
		)
		.where("accounts.id", "=", accountId)
		.where("receiptParticipants.receiptId", "=", receipt.id)
		.select("role")
		.executeTakeFirst();
	if (!participant) {
		return;
	}
	const parsed = roleSchema.safeParse(participant.role);
	if (!parsed.success) {
		// TODO: add database-level validation
		return;
	}
	return parsed.data;
};

export const getReceiptById = <SE extends ReceiptsSelectExpression<"receipts">>(
	database: Database,
	id: ReceiptsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "receipts", SE> | undefined> =>
	database
		.selectFrom("receipts")
		.select(selectExpression)
		.where("id", "=", id)
		.executeTakeFirst();

export const getOwnReceipts = (
	database: Database,
	ownerAccountId: AccountsId
) =>
	database.selectFrom("receipts").where("ownerAccountId", "=", ownerAccountId);

export const getForeignReceipts = (
	database: Database,
	ownerAccountId: AccountsId
) =>
	database
		.selectFrom("users")
		.where("users.connectedAccountId", "=", ownerAccountId)
		.where("users.ownerAccountId", "<>", ownerAccountId)
		.innerJoin("receiptParticipants", (jb) =>
			jb.onRef("receiptParticipants.userId", "=", "users.id")
		)
		.innerJoin("receipts", (jb) =>
			jb.onRef("receipts.id", "=", "receiptParticipants.receiptId")
		);
