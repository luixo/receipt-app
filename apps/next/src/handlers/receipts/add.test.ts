import { faker } from "@faker-js/faker";
import { describe, expect } from "vitest";

import { createAuthContext } from "@tests/backend/utils/context";
import {
	insertAccount,
	insertAccountWithSession,
	insertUser,
} from "@tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
	expectUnauthorizedError,
} from "@tests/backend/utils/expect";
import { test } from "@tests/backend/utils/test";
import { t } from "next-app/handlers/trpc";
import { UUID_REGEX } from "next-app/handlers/validation";

import { procedure } from "./add";
import {
	getValidReceipt,
	verifyCurrencyCode,
	verifyIssued,
	verifyName,
} from "./utils.test";

const router = t.router({ procedure });

describe("receipts.add", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			router.createCaller(context).procedure(getValidReceipt()),
		);

		verifyName(
			(context, name) =>
				router.createCaller(context).procedure({ ...getValidReceipt(), name }),
			"",
		);

		verifyCurrencyCode(
			(context, currencyCode) =>
				router
					.createCaller(context)
					.procedure({ ...getValidReceipt(), currencyCode }),
			"",
		);

		verifyIssued(
			(context, issued) =>
				router
					.createCaller(context)
					.procedure({ ...getValidReceipt(), issued }),
			"",
		);

		describe("participants", () => {
			test("invalid uuid", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				const invalidUuid = "not-a-uuid";
				await expectTRPCError(
					() =>
						caller.procedure({
							...getValidReceipt(),
							participants: [faker.string.uuid(), invalidUuid],
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "participants[1]": Invalid uuid`,
				);
			});
		});

		test("user does not exist", async ({ ctx }) => {
			const {
				sessionId,
				accountId,
				account: { email },
			} = await insertAccountWithSession(ctx);
			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await insertUser(ctx, accountId);
			const fakeUserId = faker.string.uuid();
			await expectTRPCError(
				() =>
					caller.procedure({
						...getValidReceipt(),
						participants: [fakeUserId],
					}),
				"NOT_FOUND",
				`User "${fakeUserId}" does not exist or is not owned by "${email}".`,
			);
		});

		test("user is not owned by an account", async ({ ctx }) => {
			const { sessionId, accountId, account } = await insertAccountWithSession(
				ctx,
			);
			await insertUser(ctx, accountId);

			const { id: foreignAccountId } = await insertAccount(ctx);
			const { id: foreignUserId } = await insertUser(ctx, foreignAccountId);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						...getValidReceipt(),
						participants: [foreignUserId],
					}),
				"NOT_FOUND",
				`User "${foreignUserId}" does not exist or is not owned by "${account.email}".`,
			);
		});
	});

	describe("functionality", () => {
		test("receipt is added", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);

			// Verify unrelated data doesn't affect the result
			await insertUser(ctx, accountId);
			const { id: foreignAccountId } = await insertAccount(ctx);
			await insertUser(ctx, foreignAccountId);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			const result = await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure(getValidReceipt()),
			);
			expect(result).toMatch(UUID_REGEX);
		});

		test("receipt is added - with users", async ({ ctx }) => {
			const {
				sessionId,
				accountId,
				userId: selfUserId,
			} = await insertAccountWithSession(ctx);
			const { id: userId } = await insertUser(ctx, accountId);

			// Verify unrelated data doesn't affect the result
			await insertUser(ctx, accountId);
			const { id: foreignAccountId } = await insertAccount(ctx);
			await insertUser(ctx, foreignAccountId);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			const result = await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure({
					...getValidReceipt(),
					participants: [userId, selfUserId],
				}),
			);
			expect(result).toMatch(UUID_REGEX);
		});
	});
});
