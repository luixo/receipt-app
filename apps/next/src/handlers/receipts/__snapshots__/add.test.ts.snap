// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`receipts.add > functionality > receipt is added - with users 1`] = `
{
  "added": {
    "receiptParticipants": {
      "receiptId:23345cd5-9763-4f6c-950c-0d92511f1e71|userId:19dc7ddc-bbdb-4896-bd92-2deed2136929": {
        "added": "2020-01-01T00:00:00.000Z",
        "receiptId": "23345cd5-9763-4f6c-950c-0d92511f1e71",
        "resolved": false,
        "role": "owner",
        "userId": "19dc7ddc-bbdb-4896-bd92-2deed2136929",
      },
      "receiptId:23345cd5-9763-4f6c-950c-0d92511f1e71|userId:be70a9a6-7d78-45ed-ba31-8b5daefdb497": {
        "added": "2020-01-01T00:00:00.000Z",
        "receiptId": "23345cd5-9763-4f6c-950c-0d92511f1e71",
        "resolved": false,
        "role": "editor",
        "userId": "be70a9a6-7d78-45ed-ba31-8b5daefdb497",
      },
    },
    "receipts": {
      "id:23345cd5-9763-4f6c-950c-0d92511f1e71": {
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "SGD",
        "id": "23345cd5-9763-4f6c-950c-0d92511f1e71",
        "issued": "2020-01-01T00:00:00.000Z",
        "lockedTimestamp": null,
        "name": "consequatur illo eius",
        "ownerAccountId": "19dc7ddc-bbdb-4896-bd92-2deed2136929",
      },
    },
  },
}
`;

exports[`receipts.add > functionality > receipt is added 1`] = `
{
  "added": {
    "receipts": {
      "id:2d658309-93f9-4fee-bd0f-a1abdb942283": {
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "GNF",
        "id": "2d658309-93f9-4fee-bd0f-a1abdb942283",
        "issued": "2020-01-01T00:00:00.000Z",
        "lockedTimestamp": null,
        "name": "commodi vitae error",
        "ownerAccountId": "ba8bebfc-741c-49a8-b07f-2a80c4e34146",
      },
    },
  },
}
`;
